### ISSUE TITLE
(De)activation of plugins doesn't re-execute plugin hooks correctly

### ISSUE THREAD

**Issue by webketje** [GetSimpleCMS/GetSimpleCMS#1116]
In GS 3.3.7 I wanted to hide tabs through the header/ footer hooks, based on some conditions.
The (de)activation of plugins happens through AJAX, and somehow plugin hooks are not loaded correctly, when the header/ sidebar/ main content is received. This stuff has to do with the file `jquery.getsimple.js`.

**Comment by tablatronix** [comment]
Yeah I have never bothered looking into this, as it is very rare issue.

If you have a reproduction scenario, please give me a testbed example so I can fix it, also see if the issue exists in master branch already.

**Comment by tablatronix** [comment]
Is this a condition of toggling any plugin , or the one you are actually trying to use ?
eg. does a refresh "fix" the problem

**Comment by webketje** [comment]
Actually I wanted to create a new MultiUser plugin (based on the concept of user groups, eg. editor, manager, admin, front-end etc.), and the only case where I was unable to permanently hide/remove unauthorized tabs was here. 

Here's a sample plugin you can use to test it ([1] have some other plugins installed, [2] activate this one, and reload: only Pages, Theme, and Plugins tab should be visible, [3] (de)activate another plugin => all tabs are visible again: 

``````````
<?php

// register plugin
register_plugin(
	'plugins_reload_bug',
	'Plugins reload bug',
	'1',
	'Kevin Van Lierde',
	'http://gsbug.dummy', 
	'This plugin demonstrates the incorrect plugin hook re-execution on the plugin page after (de)activating a plugin.',
	'',
  ''
);

add_action('header', 'plugins_reload_bug_hook');

$perms = array(
	'access_theme_tab',
	'access_pages_tab',
	'access_plugins_tab'
);

function plugins_reload_bug_hook() {
	global $perms;
	echo '<style id="plugin_reload_style_test">';
	foreach($perms as $perm) {
		if (user_cannot($perm))
			echo str_replace('access', '#nav', str_replace('_tab','', $perm)) . ' {display: none;} '; 
	}
	echo '</style>
	<script id="plugin_reload_script_test">
		jQuery(document).ready(function() { 
		  var array = ' . json_encode($perms) . ';
			$.each($(\'[id^=nav_]\'), function(index, tab) {
				if (array.indexOf(\'access\' + tab.id.replace(\'nav\',\'\') +\'_tab\') === -1)
					$(tab).remove();
			});
		});
	</script>';
}

function user_cannot($permission)
{
	global $perms;
	if (array_search($permission, $perms) > -1)
		return false;
	return true;
}
``````

**Comment by tablatronix** [comment]
You should rad my notes on that old plugin I had a pretty good idea on implementing a standard pattern for adding custom permissions ( hacks )

I proposed a mapping record for a filter permission
- [ ] page id
- [ ] action / qs
- [ ] jquery target
- [ ] POST
- [ ] GET

etc. the record wouldh have an ID, a desc, and these standard pattern that would fire.

**Comment by tablatronix** [comment]
plugins uses an old ajax method to refresh pieces of the page.

It loads plugins.php, then selectively replaces
header
sidebar
maincontent
But it first filters all script, so that is is not executed, which results in assets and stuff downloading and all kinds of weird shit.

Not sure how i can fix this without looking further into it and redesigning this.
Maybe for 3.4 i will.

You might be able to use a delegate or mutate watcher to refire your functions or a simple global ajax.success() bind with a slight delay perhaps.

**Comment by webketje** [comment]
The filter permission looks like a nice idea. The plugin I wanted to create (in a first phase) was a bare-bones user API (as I'm more interested in data structures than implementations), which would provide the system to register, load and manipulate permissions, users, user properties, and user groups, all without touching the original user XML files. 

As for the bug: 
>When the event bound to .toggleEnable is unbound with .die('click') all works as expected.

Without AJAX, the plugin is still updated, however the update notice is not displayed (although this could be done in `plugins.php` easily). So quickest way out, for what it's worth, would be to let PHP handle the request.

**Comment by tablatronix** [comment]
Yeah that is what i proposed, the permissions plugin was lacking any organization or api to easily add blocking patterns to it which was very annoying.

I doubt i will be removing ajax from something while in 3.4 the goal was to make more stuff ajax, and eventually the entire UI. I can fix the update notice part, that should probably be working without js anyway. Some pages ajax can be toggled on or off via a body class in 3.4 eventually this will be well implemented.

There is nothing wrong without how we do it. Your plugin needs to be able to deal with dom changes.

**Comment by tablatronix** [comment]
For future features of multi user and permissions hooks, please read and contribute to the following feature discussions, These are already in 3.4 but very crudely and mostly as a proof of concept to hash out it working. 

#860 
#408

**Comment by tablatronix** [comment]
Here is an example of what i was thinking

```php
$block = array(
 pageid => array('theme.php'),
 jquery => '$('#nav_theme').hide()',
// POST   => array('pageid' => 'theme.php','action' => 'submitted'),
 CSS    => "#nav_theme {visibility:none},
)
```
